src/coreclr/jit/optimizer.cpp

```diff
@@ -7608,17 +7608,805 @@ bool Compiler::optIsRangeCheckRemovable(GenTree* tree)
+//
+// Return:
+//      True if it meets type conditions and cost conditions.	Else false.
+//
+bool OptBoolsDsc::optOptimizeBoolsChkTypeCostCond(Compiler* comp)
+{
+    noway_assert(t1.compTree->OperIs(GT_EQ, GT_NE) && t1.compTree->AsOp()->gtOp1 == c1);
+    noway_assert(t2.compTree->OperIs(GT_EQ, GT_NE) && t2.compTree->AsOp()->gtOp1 == c2);
+
+    //
+    // Leave out floats where the bit-representation is more complicated
+    // - there are two representations for 0.
+    //
+    if (varTypeIsFloating(c1->TypeGet()) || varTypeIsFloating(c2->TypeGet()))
+    {
+        return false;
+    }
+
+    // Make sure the types involved are of the same sizes
+    if (genTypeSize(c1->TypeGet()) != genTypeSize(c2->TypeGet()))
+    {
+        return false;
+    }
+    if (genTypeSize(t1.compTree->TypeGet()) != genTypeSize(t2.compTree->TypeGet()))
+    {
+        return false;
+    }
+#ifdef TARGET_ARMARCH
+    // Skip the small operand which we cannot encode.
+    if (varTypeIsSmall(c1->TypeGet()))
+        return false;
+#endif
+    // The second condition must not contain side effects
+
+    if (c2->gtFlags & GTF_GLOB_EFFECT)
+    {
+        return false;
+    }
+
+    // The second condition must not be too expensive
+
+    comp->gtPrepareCost(c2);
+
+    if (c2->GetCostEx() > 12)
+    {
+        return false;
+    }
+
+    return true;
+}
+
+//
+// optOptimizeBoolsUpdateTrees: Fold the trees based on fold type and comparison type,
+//                              update the edges, unlink removed blocks and update loop table
+//
+// Arguments:
+//      comp    Instance of Compiler class
+//
+// Notes:
+//      The pOptBoolsDsc has below key members:
+//          b1:         The first Basic Block
+//          b2:         The second Basic Block
+//          b3:         The thrid Basic Block. null if both b1 and b2 are BBJ_COND
+//          t1:         The test info for the first test tree
+//          t2:         The second info for the second test tree
+//          foldOp:     The fold operator of GT_AND or GT_OR
+//          foldType:   The type of the folded tree
+//          cmpOp:      The comparison operator of GT_EQ or GT_NE
+//
+void OptBoolsDsc::optOptimizeBoolsUpdateTrees(Compiler* comp)
+{
+    assert(b1 != nullptr && b2 != nullptr);
+
+    bool optReturnBlock = false;
+    if (b3 != nullptr)
+    {
+        optReturnBlock = true;
+    }
+
+    assert(foldOp != NULL && foldType != NULL && c1 != nullptr && c2 != nullptr);
+
+    GenTree* cmpOp1 = comp->gtNewOperNode(foldOp, foldType, c1, c2);
+    if (t1.isBool && t2.isBool)
+    {
+        // When we 'OR'/'AND' two booleans, the result is boolean as well
+        cmpOp1->gtFlags |= GTF_BOOLEAN;
+    }
+
+    GenTree* t1Comp = t1.compTree;
+    t1Comp->SetOper(cmpOp);
+    t1Comp->AsOp()->gtOp1         = cmpOp1;
+    t1Comp->AsOp()->gtOp2->gtType = foldType; // Could have been varTypeIsGC()
+    if (optReturnBlock)
+    {
+        // Update tree when b1: BBJ_COND and b2: GT_RETURN (BBJ_RETURN)
```

Comment:
... and b3: BBJ_RETURN?

You are merging 2 returns into one. You need to add:
```
assert(comp->fgReturnCount >= 2);
--comp->fgReturnCount;
```

I expect that in the case we have more than the max number of returns, and create a single return block, this optimization won't kick in.

Reply:
Done for adding code.

> I expect that in the case we have more than the max number of returns, and create a single return block, this optimization won't kick in.

TODO: The current code optimizes 2 returns. I need to know the pattern of >2 returns.

Reply:
Created https://github.com/dotnet/runtime/issues/55605 and added this item there.